### PR TITLE
NVIDIA: provide alternative drivers

### PR DESCRIPTION
**Pull request by AndWac**
This is based on Mno-himes work #5337.
I am not sure what other things are necessary:
- do we need to obsolete the original nvidia package?

### CODE REVIEW

**Comment by Toasterson** [comment]
Am I right in the assumption, that we have no more driver/graphics/nvidia right now? Then we need to make one that replaces driver/graphics/nvidia. (rename) or make driver/graphics/nvidia a metapackage, that holds the mediator links for ogl-select service. and the driver/graphics/nvidia-340 and driver/graphics/nvidia-390 have only the files themselves.

**Comment by AndWac** [comment]
Yes, driver/graphics/nvidia will be replaced by driver/graphics/nvidia-340 (for the moment). The new nvidia-390 is added. Later we can add more drivers and change the default.

**Comment by Toasterson** [comment]
Then we need to rename the old package. Meaning publish a new version with the rename action

**Comment by AndWac** [comment]
Like this? Can you check the version information, I am not sure whether to increment it or not.

**Comment by Toasterson** [comment]
I have absolutely no clue how that process works. I just know there needs to be a package driver/graphics/nvidia with a higher version number than before that has a rename action in it. (Special kind of set action) if we generate that I do not know how.

File: components/openindiana/nvidia-390/driver-graphics-nvidia.p5m [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+# A full copy of the text of the CDDL should have accompanied this
+# source.  A copy of the CDDL is also available via the Internet at
+# http://www.illumos.org/license/CDDL.
+#
+
+#
+# Copyright 2019 Michal Nowak
+#
+
+set name=pkg.fmri value=pkg:/driver/graphics/nvidia@$(IPS_COMPONENT_VERSION),$(BUILD_VERSION)
```
Comment (pyhalov) [inline components/openindiana/nvidia-390/driver-graphics-nvidia.p5m:16]: We publish two driver/graphics/nvidia packages as result.

**Comment by pyhalov** [comment]
Can we make these two drivers coexist? Look at https://github.com/OpenIndiana/oi-userland/pull/3673

**Comment by pyhalov** [comment]
While #3673 is hard to support, some simpler mediated  solution would be interesting.

File: components/openindiana/nvidia-390/driver-graphics-nvidia.p5m (hunk 1 above)
Comment (AndWac) [reply components/openindiana/nvidia-390/driver-graphics-nvidia.p5m:16]: What would be the better solution:
- keep this file and remove the history and server_desktop change
or
- remove this manifest and control explicitly?

**Comment by AndWac** [comment]
Yes, my plan is to get a simple solution now, gain some experience, maybe add more driver versions and then try to create something more comfortable.

**Comment by despair86** [comment]
ha
i ran into a nasty set of bugs with the stock installation package cause it _really_ wants to use `gfx_private` from Oct 2007

they still occur but at least when X crashes i can go back to the console (the old driver would leave me at a black screen)

unfortunately i have a fairly recent gpu which requires quadro R435+

**Comment by despair86** [comment]
re: driver errors - i was able to dig up the kmod sources from the other half of the root pkg that isn't installed - comes with the object code for the driver itself + source for the Solaris kernel interfaces - same way the linux and freebsd driver sources are shipped - this object code is _exactly the same_ for all three platforms (!)

was able to recompile the source and link against the closed-source bits (using a similar process to the way the nvidia linux kmod is rebuilt after kernel update) but the new driver triggered a page fault and was unusable

compiling with gcc-9 triggered a million and one relocation errors in krtld (gas or /usr/ccs/bin/as from a 2018 release not the 2009 version)

**Comment by jpandre** [comment]
Not sure this is related, I suspect it could be.

On bootup I see errors related to loading nvidia_modeset. As I do not have any Nvidia graphic card, I wanted to uninstall the nvidia packages, but when I uninstall nvidia-390, nvidia-340 gets installed, and conversely (on a system updated from an older version).

Is nvidia a compulsory package irrespective of Nvidia hardware ?

**Comment by Toasterson** [comment]
Hey @jpandre Merged PR's are not a really good place to ask questions, as not everybody will see it. Can you ask on the mailing list or in the discussion parts of Github?

**Pull request merged**